src/aranges.rs

```diff
@@ -0,0 +1,249 @@
+pub struct Arange
+{
+    segment: u64,
+    offset: u64,
+    length: u64,
+    header: Rc<ArangeHeader>,
+}
+
+impl Arange
+{
+    /// Parse an arange set header. Returns a tuple of the remaining arange sets, the aranges to be
+    /// parsed for this set, and the newly created ArangeHeader struct.
+    fn parse_header<Endian>(input: EndianBuf<Endian>)
+                            -> ParseResult<(EndianBuf<Endian>,
+                                            EndianBuf<Endian>,
+                                            Rc<ArangeHeader>)>
+        where Endian: Endianity
+    {
+        let (rest, (length, format)) = try!(parse_unit_length(input.into()));
+        let (rest, version) = try!(parse_u16(rest.into()));
+
+        if version != 2 {
+            return Err(Error::UnknownVersion)
+        }
+
+        let (rest, offset) = try!(parse_debug_info_offset(rest.into(), format));
+        let (rest, address_size) = try!(parse_address_size(rest.into()));
+        let (rest, segment_size) = try!(parse_address_size(rest.into()));
+
+        let header_length = match format {
+            Format::Dwarf32 => 8,
+            Format::Dwarf64 => 12,
+        };
+        let dividing_line: usize = try!(length.checked_sub(header_length).ok_or(Error::BadLength)) as usize;
+
+        Ok((rest.range_from(dividing_line..),
+            rest.range_to(..dividing_line),
+            Rc::new(ArangeHeader { format: format,
+                                   length: length,
+                                   version: version,
+                                   offset: offset,
+                                   address_size: address_size,
+                                   segment_size: segment_size } )))
+    }
+
+    /// Parse a single arange. Return `None` for the null arange, `Some` for an actual arange.
+    fn parse_range<'input, Endian>(input: EndianBuf<'input, Endian>, header: &Rc<ArangeHeader>)
+                                   -> ParseResult<(EndianBuf<'input, Endian>, Option<Arange>)>
+        where Endian: Endianity
+    {
+        let address_size = header.address_size;
+        let segment_size = header.segment_size; // May be zero!
+
+        let (rest, segment) = try!(parse_uN_as_u64(segment_size, input));
+        let (rest, offset) = try!(parse_uN_as_u64(address_size, rest));
+        let (rest, length) = try!(parse_uN_as_u64(address_size, rest));
+
+        Ok((rest, match (segment, offset, length) {
+            (0, 0, 0) => None,
+            _ => Some(Arange { segment: segment,
+                               offset: offset,
+                               length: length,
+                               header: header.clone() } ),
+        }))
+    }
+
+    /// Parse a single set of aranges, including a header and all the aranges specified. Returns
+    /// the remaining arange sets and the series of parsed aranges.
+    fn parse<Endian>(input: EndianBuf<Endian>) -> ParseResult<(EndianBuf<Endian>, Vec<Arange>)>
```

Comment:
Indeed things change if you want to symbolicate many addresses. This is why I prefer the library leave these kinds of decisions about whether to stop in the middle after finding a specific address range or building up a query-able data structure to the consumer of the library.

I'm not too worried about the function call for each 16 byte segment: the `next_arange` method can be marked `#[inline]` and there shouldn't be much penalty.

On the other hand, if you want to come back with bench numbers, I'm happy to defer to them.